File: LeetCode/1282. Group the People Given the Group Size They Belong To/solution.py
"""
103 / 103 test cases passed.
Runtime: 44 ms
Memory Usage: 15 MB
"""
class Solution:
    def groupThePeople(self, groupSizes: List[int]) -> List[List[int]]:
        rec = collections.defaultdict(list)
        for i, gs in enumerate(groupSizes):
            if not rec[gs] or len(rec[gs][-1]) == gs:
                rec[gs].append([])
            rec[gs][-1].append(i)
        return [v for value in rec.values() for v in value]

File: LeetCode/1282. Group the People Given the Group Size They Belong To/Solution.java
/**
 * 103 / 103 test cases passed.
 * Runtime: 8 ms
 * Memory Usage: 42.2 MB 
 */
class Solution {
    public List<List<Integer>> groupThePeople(int[] groupSizes) {
        HashMap<Integer, LinkedList<LinkedList<Integer>>> rec = new HashMap<>();
        for (int i = 0; i < groupSizes.length; i ++ ){
            int gs = groupSizes[i];
            if (!rec.containsKey(gs)) {
                LinkedList<LinkedList<Integer>> tmp = new LinkedList<>();
                tmp.add(new LinkedList<Integer>());
                rec.put(gs, tmp);
            } else if (rec.get(gs).getLast().size() == gs) {
                rec.get(gs).add(new LinkedList<Integer>());
            }
            rec.get(gs).getLast().add(i);
        }
        List<List<Integer>> ans = new LinkedList<>();
        for (LinkedList<LinkedList<Integer>> values: rec.values()) {
            for (LinkedList<Integer> value: values) {
                ans.add(value);
            }
        }
        return ans;
    }
}
